### ISSUE TITLE
Feature request: continuation (then) which does not require the callable object to take a future<R> as parameter

### ISSUE THREAD

**Issue by shoshijak** [STEllAR-GROUP/hpx#2591]
In the current implementation of `.then(...)`, the callable object passed as argument to a `.then(...) `has to accept a `future<R>` as argument. 

This requirement means that one can be led to writing code which looks like: 

```
	hpx::future<void> setup_future = hpx::async(
            [](){
                setup();
	  });

	hpx::future<void> computation_based_on_setup_future = setup.then(
            [](hpx::future<void> fv)
            {
	            computation();
	  });
```

Whereby the argument `hpx::future<R> fv` in the second lambda is only there to make the program compile, but does not have any meaning or real use. 

There are situations where one might want to attach a continuation B to a task A, but where B does not require any result from A, but rather just requires A to have completed. In such situations, it would be more elegant to be able to use a `.then(...)` which does not require B to accept `future<R>` as parameter.

**Comment by hkaiser** [comment]
> Whereby the argument hpx::future<R> fv in the second lambda is only there to make the program compile, but does not have any meaning or real use.

The rationale for having the continuation accept the future itself is twofold:

- the original future which was used to attach the continuation is invalidated by this operation and the only way to access it afterwards is through the future instance which is passed to the continuation itself
- a future can have two 'ready'-states: it can be ready (success) and exceptional (something went wrong), so passing the future to the continuation is the only way to access the exception (if there was any).

The bottom line is that you should always call `future::get()` on the future which is passed to the continuation as this re-throws any exceptions and allows for them to be propagated back to the user.

The use case you're referring to, e.g. attaching a continuation which does not need to refer to the result of the previous step, can be written as:

    void func_A() { ... }

    future<void> f = ...;
    future<void> f2 = f.then(hpx::util::unwrapped(&func_A));

in which case `unwrapped` will call `future::get()` on your behalf and all exceptions thrown will automatically propagate to `f2`.

**Comment by shoshijak** [comment]
Your explanation is enlightening, thank you.
